### PR TITLE
Allow compute_iou to take boxes of different shape.

### PR DESCRIPTION
**Pull request by tanzhenyu**
Allow _RPNLabelEncoder to take batched inputs.
Move the data labeling from data pipeline to model pipeline. This
 1) _might_ come with a performance cost (but I haven't noticed any regression), but 
 2) in the long run it get us better performance if we want more data augmentation in the data pipeline.
 3) make supporting `model.fit` more native and users don't have to encode labels if they build their own dataset.

### CODE REVIEW

**Comment by tanzhenyu** [comment]
It looks like many KPL are using dimension 5 instead of 4 and compute_iou should support it for now?

File: keras_cv/bounding_box/iou.py [hunk 1]
```diff
@@ -50,16 +95,19 @@ def compute_iou(
     boxes1_rank = len(boxes1.shape)
     boxes2_rank = len(boxes2.shape)
 
-    if (
-        boxes1_rank != boxes2_rank
-        or boxes1_rank not in [2, 3]
-        or boxes2_rank not in [2, 3]
-    ):
+    if boxes1_rank not in [2, 3]:
```
Comment (LukeWood) [inline keras_cv/bounding_box/iou.py:98]: can we update the docstring to add more information about the updated input/output shapes?

Comment (LukeWood) [inline keras_cv/bounding_box/iou.py:98]: perhaps instead of supporting differing ranks we should follow the TensorFlow convention and broadcast only when the first axis is of dim [1].

So instead of:

[boxes, 4], [batch, boxes, 4]

you would pass:

[1, boxes, 4], [batch, boxes, 4]

This is more consistent with the TensorFlow API than automatically broadcasting for users

Comment (tanzhenyu) [reply keras_cv/bounding_box/iou.py:98]: Here the problem is not on [1, boxes, 4] vs [batch, boxes, 4], it's actually [boxes_1, 4] & [batch, boxes_2, 4] should give user [batch, boxes_1, boxes_2, 4], if users pass in [1, boxes_1, 4] and [batch, boxes_2, 4], this will actually error out telling user that the shape doesn't match, which is expected.
If we do ask users to convert to [batch, boxes_1, 4] before calling compute_iou, then users would have to do tf.tile(tf.expand_dims(boxes1, 0), tf.stack([tf.shape(boxes_2)[0], tf.shape(boxes_1)[1:])) in order to call it, which to me is too bloated.

File: keras_cv/models/object_detection/faster_rcnn.py [hunk 2]
```diff
@@ -329,6 +329,47 @@ def __init__(
...
+            rcnn_box_pred = _decode_deltas_to_boxes(
+                anchors=rois,
+                boxes_delta=rcnn_box_pred,
+                anchor_format="yxyx",
+                box_format=self.bounding_box_format,
+                variance=[0.1, 0.1, 0.2, 0.2],
+            )
+        return rcnn_box_pred, rcnn_cls_pred
+
+    # def train_step(self, data):
```
Comment (ianstenbit) [inline keras_cv/models/object_detection/faster_rcnn.py:370]: nit: looks like this got left behind

File: keras_cv/bounding_box/iou.py (hunk 1 above)
Comment (tanzhenyu) [reply keras_cv/bounding_box/iou.py:103]: Done.

Comment (tanzhenyu) [reply keras_cv/bounding_box/iou.py:103]: I have added 4 different scenarios on the documentation part so hopefully this will give user more context

Comment (ianstenbit) [reply keras_cv/bounding_box/iou.py:103]: I'm no expert in the bbox conventions, but personally I think this API looks good.

**Review by ianstenbit — APPROVED** [review]

File: keras_cv/bounding_box/iou.py (hunk 1 above)
Comment (LukeWood) [reply keras_cv/bounding_box/iou.py:103]: Sounds good, that’s fine with me.

**Review by LukeWood — APPROVED** [review]

File: keras_cv/models/object_detection/faster_rcnn.py (hunk 2 above)
Comment (tanzhenyu) [reply keras_cv/models/object_detection/faster_rcnn.py]: Oops..removed
